Add support for Nordic's nRF52810 IC

carlescufi opened the pull request:
Adds support for the IC itself and a new board which enforces the 52810 limitations but mirrors the nRF52 DK hardware, since there is no development kit specific to the 52810.

Tested with the RTT console since UART is not functional yet.

Note that BLE is not functional yet either. We require a new version of nrfx with the appropriate radio bindings and then @cvinayak will be able to add a layer for it in the bluetooth hal folder.

carlescufi committed:
arch: arm: nrf: Add support for the nRF52810

The nRF52810 is a low-cost variant of the nRF52832, with a reduced set
of peripherals and memory. This commit adds basic support for it in the
arch SoC and dts folders.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Ioannis Glaropoulos <ioannis.glaropoulos@nordicsemi.no>

carlescufi committed:
boards: arm: Add Nordic nRF52810 board

Since Nordic does not offer an nRF52810 Development Kit and instead
recommends using the nRF52 DK (nRF52832 based), add a board that
enforces the limitations of the nRF52810 but is designed to be flashed
onto the actual nRF52 DK hardware.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Ioannis Glaropoulos <ioannisg.glaropoulos@nordicsemi.no>

carlescufi committed:
ext: hal: nordic: nrfx: Add nRF52810 configuration

nrfx requires specific chip configuration for each IC. Those are not
upstream in the nrfx repository, but rather part of the Zephyr nrfx
port. Add a configuration for the nRF52810.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Ioannis Glaropoulos <ioannis.glaropoulos@nordicsemi.no>

carlescufi commented on include/arch/arm/cortex_m/mpu/arm_mpu.h:
```diff
@@ -155,6 +155,8 @@ struct arm_mpu {
 #define REGION_1G       (0x1D << 1)
 #define REGION_2G       (0x1E << 1)
 #define REGION_4G       (0x1F << 1)
+#define REGION_24K      (0x20 << 1)
```
@galak @agross-linaro @ioannisg please review this change

ioannisg replied:
This is not correct. These macros are used to encode the region size in the MPU register, when configuring MPU regions, and AFAIK, we only have powers of two, there.
Check table B3-43 in ARMv7-M technical reference manual.

carlescufi commented:
CC @junqingzou @jarz-nordic @kl-cruz

ioannisg commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA:
```diff
@@ -0,0 +1,20 @@
+# Kconfig.defconfig.nrf52810 - Nordic Semiconductor nRF52832 MCU
```
Fix number

carlescufi replied:
fixed

jarz-nordic commented on arch/arm/soc/nordic_nrf/nrf52/soc.c:
```diff
@@ -381,7 +381,7 @@ static int nordicsemi_nrf52_init(struct device *arg)
 
-#ifdef CONFIG_SOC_NRF52832
+#if defined(CONFIG_SOC_NRF52832) || defined(CONFIG_SOC_NRF52810)
 	nordicsemi_nrf52832_init();
```
I think that both chips shall have different init functions. Most of PANs were fixed in nrf52810

ioannisg replied:
correct.

carlescufi replied:
Fixed

ioannisg commented on boards/arm/nrf52810_pca10040/Kconfig:
```diff
@@ -0,0 +1,9 @@
+# Kconfig - nRF52810 PCA10040 board configuration
```
pca10068 - also correct folder name

carlescufi replied:
No, see commit messages for explanation

ioannisg replied:
ok.

ioannisg commented on boards/arm/nrf52810_pca10040/Kconfig:
```diff
@@ -0,0 +1,9 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_NRF52810_PCA10040
```
here too, and everywhere else

carlescufi replied:
same

ioannisg replied:
By the way @carlescufi do we really need these empty Kconfig files?

jarz-nordic commented on boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts:
```diff
@@ -0,0 +1,74 @@
+		     "nordic,nrf52810";
+
+	chosen {
+		zephyr,console = &uart0;
```
nrf52810 does not have an UART0, only UARTE0

galak replied:
While @carlescufi comment is correct, if its the case that UARTE0 can only be the 'E' form, than the compat in the SOC dts should just be 'nordic,nrf-uarte' and we don't need to update the compat in the board.dts

carlescufi replied:
It's already the case that compat in the SoC dts is only nrf-uarte

ioannisg replied:
then i suppose we are ok.

jarz-nordic commented on dts/arm/nordic/nrf52810.dtsi:
```diff
@@ -0,0 +1,83 @@
+			label = "I2C_0";
+		};
+
+		i2c1: i2c@40004000 {
```
nrf52810 does not have TWIM1.
It has only TWIM0 at address 40003000

carlescufi replied:
fixed

nvlsianpu commented on boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts:
```diff
@@ -0,0 +1,74 @@
+			label = "image-1";
+			reg = <0x00016000 0xa000>;
+		};
+		scratch_partition: partition@20000 {
```
Why the scratch partition is of the image slot size? The size of 0x2000 will be better here (1/5 so, the same ratio as for nrf52832).

carlescufi replied:
Fixed, thanks

galak commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.series:
```diff
@@ -8,7 +8,7 @@
 	bool "Nordic Semiconductor nRF52 series MCU"
 	select CPU_CORTEX_M4
-	select CPU_HAS_FPU
+	select CPU_HAS_FPU if !SOC_NRF52810
```
At this point, seems like move the 'select CPU_HAS_FPU' over to the SOC_NRF52... would be better.

carlescufi replied:
Done

galak commented on boards/arm/nrf52810_pca10040/board.h:
```diff
@@ -0,0 +1,48 @@
+#include <soc.h>
+
+/* Push button switch 0 */
+#define SW0_GPIO_PIN     13
```
Can we do this via DTS instead since this is a new board.

carlescufi replied:
I would love to, but I think I would need some help here. Can you give me some pointers?

galak replied:
You'll also need to pull in some changes from PR #7444 (meaning adding the dt nodes for GPIO etc).

carlescufi replied:
Very well. In that case I will wait until #7444 is merged, since this should not be merged before 1.12 is released anyway. Thanks for the pointers.

galak commented on boards/arm/nrf52810_pca10040/nrf52810_pca10040.dts:
```diff
@@ -0,0 +1,74 @@
+
+&uart0 {
+	status = "ok";
+	compatible = "nordic,nrf-uarte";
```
See above comment if this compat should just be in the SoC dtsi

carlescufi replied:
I removed this line from the board dts file, since it's already in the dtsi

galak reviewed (dismissed):
Looks good, a few dts and other style comments.

carlescufi committed:
boards: 96_nitrogen: Specify the IC

In order to avoid auto-selection by Kconfig, the board needs to define
the specific IC that it contains, in this case an nRF52832.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
tests: Fix sizing for several test for chips with 24KB of RAM

When adding the nRF52810, which has 24KB of RAM, some of the tests don't
compile anymore due to lack of SRAM. Address this by either filtering
the test out or reducing the amount of memory allocation.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+// <e> NRFX_SPI_ENABLED - nrfx_spi - SPI peripheral driver
+//==========================================================
+#ifdef CONFIG_NRFX_SPI
+#define NRFX_SPI_ENABLED 1
```
nRF52810 is not equipped with SPI. There are SPIM and SPIS

carlescufi replied:
I pushed a new file, please re-review

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+
+
+#ifndef NRFX_TIMER3_ENABLED
+#define NRFX_TIMER3_ENABLED 0
```
There is no TIMER3 in nRF52810

carlescufi replied:
I pushed a new file, please re-review

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+
+
+#ifndef NRFX_TIMER4_ENABLED
+#define NRFX_TIMER4_ENABLED 0
```
There is no TIMER4 in nRF52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+// <e> NRFX_I2S_ENABLED - nrfx_i2s - I2S peripheral driver
+//==========================================================
+#ifndef NRFX_I2S_ENABLED
+#define NRFX_I2S_ENABLED 0
```
no I2S in nRF52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+// <e> NRFX_LPCOMP_ENABLED - nrfx_lpcomp - LPCOMP peripheral driver
+//==========================================================
+#ifndef NRFX_LPCOMP_ENABLED
+#define NRFX_LPCOMP_ENABLED 0
```
no LPCOMP in nRF52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+
+
+#ifndef NRFX_PWM1_ENABLED
+#define NRFX_PWM1_ENABLED 0
```
only PWM0 in nRF52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+
+
+#ifdef CONFIG_SPI_1_NRF_SPIM
+#define NRFX_SPIM1_ENABLED 1
```
only SPIM0 in 52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+
+
+#ifndef NRFX_TWIM1_ENABLED
+#define NRFX_TWIM1_ENABLED 0
```
only TWIM0 in nRF52810

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2900 @@
+// <e> NRFX_TWI_ENABLED - nrfx_twi - TWI peripheral driver
+//==========================================================
+#ifndef NRFX_TWI_ENABLED
+#define NRFX_TWI_ENABLED 0
```
no TWI in nRF52810.
This config file needs to be changed

kl-cruz reviewed (changes requested):
config file needs rework

pizi-nordic commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -5,18 +5,29 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
+config SOC_NRF52810
```
There is one more difference between nRF52810 and other members of the nRF52 family: The nRF52810 does not have instruction cache, which should be enabled on other SoCs.

At the moment it looks that we are not enabling cache at all. IMHO we should fix that asap as without cache performance is significantly degraded. Please fix that in this PR or create new issue.

carlescufi commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.defconfig.nrf52810_QFAA:
```diff
@@ -0,0 +1,20 @@
+
+config SOC
+	string
+	default nRF52810_QFAA
```
Fixed, thanks

ulfalizer commented on boards/arm/nrf52810_pca10040/Kconfig.defconfig:
```diff
@@ -0,0 +1,35 @@
+if BOARD_NRF52810_PCA10040
+
+config BOARD
+	default nrf52810_pca10040
```
Still missing quotes here.

carlescufi replied:
Fixed

ulfalizer reviewed (approved):
Quoting nit addressed

ioannisg commented on arch/arm/soc/nordic_nrf/nrf52/mpu_regions.c:
```diff
@@ -18,10 +18,21 @@ static struct arm_mpu_region mpu_regions[] = {
 	MPU_REGION_ENTRY("FLASH_0",
 			 CONFIG_FLASH_BASE_ADDRESS,
 			 REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
+#ifdef REGION_FLASH_1_SIZE
```
use #if defined(.....)

ioannisg replied:
Fixed.

ioannisg commented:
@carlescufi MPU regions' configuration looks good, now. 
I left some styling comments only.

anangl commented on dts/arm/nordic/nrf52810.dtsi:
```diff
@@ -0,0 +1,72 @@
+		};
+
+		i2c0: i2c@40003000 {
+			compatible = "nordic,nrf5-i2c";
```
nRF52810 has only TWIM and no TWI that is present on nRF51 and other nRF52s. Shouldn't a new binding ("nordic,nrf-twim") be introduced and this value changed?

dbkinder reviewed (changes requested):
small tweak to the doc

kl-cruz commented on ext/hal/nordic/nrfx_config_nrf52810.h:
```diff
@@ -0,0 +1,2101 @@
+
+// <e> NRFX_TWIM_ENABLED - nrfx_twim - TWIM peripheral driver
+//==========================================================
+#ifndef NRFX_TWIM_ENABLED
```
When this PR is merged (https://github.com/zephyrproject-rtos/zephyr/pull/7788) it should be changed to CONFIG_NRFX_TWIM like i the SPIM section. Line 1665 too.

ioannisg replied:
Will be addressed, thanks.

ioannisg commented on tests/kernel/sched/schedule_api/src/test_priority_scheduling.c:
```diff
@@ -8,8 +8,11 @@
+	#define NUM_THREAD 2
+#elif CONFIG_SRAM_SIZE <= 32
 	#define NUM_THREAD 3
+#elif defined(CONFIG_SOC_NRF52810)
```
Is this intended, @carlescufi ?

ioannisg commented on arch/arm/soc/nordic_nrf/nrf52/soc.c:
```diff
@@ -26,6 +26,8 @@ extern void _NmiInit(void);
 #endif
 
 #if defined(CONFIG_SOC_NRF52832)
+#include <system_nrf52810.h>
```
@anangl take a look here: I simply added nRF52810 nrf header, in soc.c.
Is there anything else, (besides defining DEVELOP_IN_NRF52832) that we need to do here?

anangl replied:
Good. But the line above should be `#if defined(CONFIG_SOC_NRF52810)`.
And that should be it.

ioannisg replied:
ups, my bad, thanks for catching this.
Will fix.

ioannisg committed:
boards: arm: nrf: add directive for nrf52810_pca10040 initialization

This commit adds a compile-time directive for the
nrf52810_pca10040 board that instruct nRFx to perform
certain board initialization. The directive is needed
as nrf52810_pca10040 mirrors the nrf52832 DK hardware.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -44,6 +44,7 @@ config SOC_NRF52810
 
 config SOC_NRF52832
 	depends on SOC_SERIES_NRF52X
+	select CPU_HAS_FPU
```
This addition is no longer needed (see 2 lines below).

ioannisg replied:
thanks.

anangl commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -107,6 +108,7 @@ config SOC_NRF52832
 
 config SOC_NRF52840
 	depends on SOC_SERIES_NRF52X
+	select CPU_HAS_FPU
```
This addition is no longer needed (see 2 lines below).

ioannisg replied:
thanks

anangl commented on boards/arm/nrf52810_pca10040/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# The nrf52810_pca10040 board mirrors the nRF52832 DK hardware, which
+# needs to be considered by certain system initialization functionality
+# residing in nRFx.
```
We used to write nrfx as nrfx. I think we should stay with this convention.
Anyway, I think it would be better to refer directly to the "system_nrf52810.c" file, instead of pointing to nrfx or MDK that resides inside it.

ioannisg commented on boards/arm/nrf52810_pca10040/Kconfig.defconfig:
```diff
@@ -0,0 +1,35 @@
+
+endif # GPIO_NRF5
+
+if UART_NRFX
```
Is this needed @anangl ? Is 810 going to have only UARTE, instead?

anangl replied:
It is needed. It enables nrfx drivers for UARTs in general. The type of the driver (UART or UARTE) is selected with a different option.

ioannisg replied:
ok

ioannisg commented on boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst:
```diff
@@ -0,0 +1,24 @@
+********
+
+The nrf52810_pca10040 board is a modified version of the :ref:`nrf52_pca10040`
+so that it enforces the limitations imposed by the nRF52810 IC, which is a
```
fixed

jarz-nordic commented on boards/arm/nrf52810_pca10040/doc/nrf52810_pca10040.rst:
```diff
@@ -0,0 +1,24 @@
+********
+
+The nrf52810_pca10040 board is a modified version of the :ref:`nrf52_pca10040`
+that enforces the limitations imposed by the nRF52810 IC, which is a
```
I am wondering if we shall mention here that this solution combines limations of nrf52832 and nrf52810.
User shall be aware that PANs fixed in 52810 are still valid with this approach, hence intialization time will be extended.
Furthermore nr52810 has 10bit transfer buffers for SPIM, TWIM, UARTE but you can use only 8 bits because it is HW limation in nr52832.

ioannisg replied:
I think this is a rather generic comment, and we should not care about the details. I think we just need to stress that 810 is a reduced version of 832.

ioannisg committed:
boards: arm: nrf: move LED and Button definitions in DTS

This commit moves the definitions for the LED and Buttons
supported in nrf52810_pca10040 DK in DTS from board.h. Aliases
are kept in board.h to make basic examples pass.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl commented on arch/arm/soc/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -223,5 +228,5 @@ config GPIO_AS_PINRESET
 config NRF_ENABLE_ICACHE
 	bool "Enable the instruction cache (I-Cache)"
-	depends on SOC_SERIES_NRF52X
+	depends on SOC_NRF52832 || SOC_NRF52840
```
The same should be done for `config NFCT_PINS_AS_GPIOS` - this option does not make sense for nRF52810.

ioannisg replied:
Will do.

ioannisg committed:
boards: arm: nrf: update I2C-related defines for nrf52810

The commit modifies the I2C-related defines in nrfx
configuration to comply with the updates in (57e3644).
It also modifies the i2c configuration in DTS for board
nrf52810_pca10040 to reflect the nRF52810 capabilities.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf: select HAS_DTS_I2C for nrf52_pca10040

This commit selects HAS_DTS_I2C in nRF8232 DK board definition.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

anangl commented:
Since #7334 is now in master, "nrfx_config_nrf52810.h" needs one more update: [this](https://github.com/zephyrproject-rtos/zephyr/pull/7334/files#diff-a56c8d031d7dadfe354891d51ab4b4f5L959) and [this](https://github.com/zephyrproject-rtos/zephyr/pull/7334/files#diff-a56c8d031d7dadfe354891d51ab4b4f5L965) change.
Hence, maybe it would make sense to merge #8525 first and then update commit f85cfced1a8b39bd5dd2335a98f322075131be4c regarding "nrfx_config_nrf52810.h" and "CMakeLists.txt"?

carlescufi commented:
@anangl addressed, thank you

carlescufi commented:
recheck

carlescufi commented:
recheck

carlescufi commented:
@dbkinder can you revisit this one?

agross-linaro commented:
looks good to me

carlescufi commented:
@galak @dbkinder will you be able to take a look at this soon? Your requests have been addressed so unless I hear back soon I will need to dismiss the reviews in order to get this merged. Thanks!

carlescufi commented:
thanks @dbkinder!
